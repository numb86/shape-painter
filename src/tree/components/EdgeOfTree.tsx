import React from 'react';
import styled from 'styled-components';

import {Edge} from '@tree/tree';
import {TREE} from '@src/constants';

type Props = Omit<Edge, 'id' | 'color'> & {color: string};

export const EdgeOfTree = (props: Props) => <EdgeStyledComponent {...props} />;

const EdgeStyledComponent = styled.div<Props>((props: Props) => {
  const {top, left, width, positionOfParent, color} = props;

  const linerGradientMap = new Map([
    ['center', 'right'],
    ['right', 'left top'],
    ['left', 'right top'],
  ]);

  const linerGradient = linerGradientMap.get(positionOfParent);

  let startEdgePoint = 49.2;
  let endEdgePoint = 50.8;

  if (linerGradient !== 'right') {
    startEdgePoint = 49;
    endEdgePoint = 51;
  }

  if (width < 5) {
    startEdgePoint = 40;
    endEdgePoint = 60;
  }

  let backgroundColor = '';
  if (width < 3) {
    backgroundColor = color;
  }

  return `
position: absolute;
top: ${top}px;
left: ${left}px;
width: ${width}px;
height: ${
    TREE.NODE.HEIGHT * TREE.NODE_COUNT_BETWEEN_ROWS + TREE.EDGE.LENGTH_TO_ADD
  }px;
background-image linear-gradient(to ${linerGradient}, rgba(255,255,255,0) ${startEdgePoint}%, ${color} ${
    startEdgePoint + 1
  }%, ${color} ${endEdgePoint - 1}%, rgba(255,255,255,0) ${endEdgePoint}%);
background-color: ${backgroundColor}
`;
});
