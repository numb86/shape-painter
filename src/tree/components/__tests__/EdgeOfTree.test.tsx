import React from 'react';
import {render} from '@testing-library/react';
import 'jest-styled-components';

import {EdgeOfTree} from '../EdgeOfTree';

describe('EdgeOfTree', () => {
  it('The content passed in props is reflected in the style', () => {
    const top = 20;
    const left = 300;
    const width = 100;
    const edge = {
      top,
      left,
      width,
      positionOfParent: 'center',
      color: 'skyblue',
    } as const;

    const {container: c1} = render(<EdgeOfTree {...edge} />);
    expect(c1.firstChild).toMatchSnapshot();

    const {container: c2} = render(
      <EdgeOfTree {...edge} positionOfParent="right" />
    );
    expect(c2.firstChild).toMatchSnapshot();

    const {container: c3} = render(
      <EdgeOfTree {...edge} positionOfParent="left" />
    );
    expect(c3.firstChild).toMatchSnapshot();
  });
});
