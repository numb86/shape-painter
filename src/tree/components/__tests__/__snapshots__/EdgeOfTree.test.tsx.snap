// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdgeOfTree The content passed in props is reflected in the style 1`] = `
.c0 {
  position: absolute;
  top: 20px;
  left: 300px;
  width: 100px;
  height: 93px;
  background-image: linear-gradient(to right,rgba(255,255,255,0) 49.2%,skyblue 50.2%,skyblue 49.8%,rgba(255,255,255,0) 50.8%);
}

<div
  class="c0"
  color="skyblue"
  width="100"
/>
`;

exports[`EdgeOfTree The content passed in props is reflected in the style 2`] = `
.c0 {
  position: absolute;
  top: 20px;
  left: 300px;
  width: 100px;
  height: 93px;
  background-image: linear-gradient(to left top,rgba(255,255,255,0) 49%,skyblue 50%,skyblue 50%,rgba(255,255,255,0) 51%);
}

<div
  class="c0"
  color="skyblue"
  width="100"
/>
`;

exports[`EdgeOfTree The content passed in props is reflected in the style 3`] = `
.c0 {
  position: absolute;
  top: 20px;
  left: 300px;
  width: 100px;
  height: 93px;
  background-image: linear-gradient(to right top,rgba(255,255,255,0) 49%,skyblue 50%,skyblue 50%,rgba(255,255,255,0) 51%);
}

<div
  class="c0"
  color="skyblue"
  width="100"
/>
`;
